extract_sets/training_sets.py

```diff
@@ -0,0 +1,398 @@
+        y_height = int(element.attrib['HEIGHT'])
+        return y_start + y_height
+
+
+class PageLine(TextLine):
+    """Extract TextLine Information from PAGE Data"""
+
+    def __init__(self, element, namespace, reorder):
+        super().__init__(element, namespace)
+        self.set_id()
+        self.set_text()
+        if self.valid:
+            self.reorder = reorder
+            self.shape = self.get_shape(self.element)
+
+    def set_id(self):
+        self.element_id = self.element.attrib['id']
+
+    def set_text(self):
+        """
+        * set words as preferred text source, otherwise use text line
+        * drop rtl-mark if contained
+        * print lines without coords
+        """
+
+        texts = []
+        text_els = self.element.findall(f'{self.namespace}:Word', XML_NS)
+        for t in text_els:
+            top_left = PageLine._pick_top_left(t, self.namespace)
+            if not top_left:
+                elem_id = t.attrib['id']
+                msg = f"Invalid Coords of Word '{elem_id}' in '{self.element_id}'!"
+                raise RuntimeError(msg)
+            texts.append(t)
+
+        # if no Word assume at least TextLine exists
+        if not text_els:
+            top_left = PageLine._pick_top_left(self.element, self.namespace)
+            if not top_left:
+                elem_id = self.element.attrib['id']
+                print("[ERROR  ] skip '{}': invalid coords!".format(
+                    elem_id), file=sys.stderr)
+                self.valid = False
+                return
+            texts.append(self.element)
+
+        sorted_els = sorted(
+            texts,
+            key=lambda w: int(PageLine._pick_top_left(w, self.namespace)))
+        unicodes = [
+            w.find(
+                f'.//{self.namespace}:Unicode',
+                XML_NS) for w in sorted_els]
+        self.text_words = [u.text.strip() for u in unicodes if u.text]
+
+        # elimiate read order mark
+        for i, strip in enumerate(self.text_words):
+            strip = self.text_words[i]
+            for mark in CLEAR_MARKS:
+                if mark in strip:
+                    self.text_words[i] = strip.replace(mark, '')
+
+    @staticmethod
+    def _pick_top_left(elem, namespace):
+        coords = elem.find(f'{namespace}:Coords', XML_NS)
+        points = coords.attrib['points'].split()
+        if len(points) > 0:
+            return points[0].split(',')[0]
+
+    def get_shape(self, element):
+        """
+        Coordinate data from current OCR-D-Workflows can contain
+        lots of points, therefore additional calculations are required
+        """
+
+        p_attr = element.find(
+            f'{self.namespace}:Coords',
+            XML_NS).attrib['points']
+        numbers = [int(n) for pair in p_attr.split() for n in pair.split(',')]
+
+        # group clustering idiom
+        points = list(zip(*[iter(numbers)] * 2))
+
+        return Polygon(points)
```

Comment:
That does not make sense to me. Why do you ungroup the points, only to regroup them right away?

```suggestion
        return Polygon(tuple(map(int, pair.split(','))) for pair in p_attr.split())
```

Reply:
Actually, I did it this way to make clear that there are two distinct actions to take: first parse the textual information from the coords attribute, second generate a list of int pairs that finally form the points. To separate them doesn't look as elegant as your suggestion, but it's easier to understand.

Reply:
I see. Point taken that the one-liner is not as accessible, but ungrouping is not necessary, so that's not very understandable either IMHO. I was led to believe there's more going on initially, because of the docstring says
> Coordinate data from current OCR-D-Workflows can contain lots of points, therefore additional calculations are required

So this was probably more a TODO than a description of the current implementation? (You could use Shapely's `Polygon.simplify(tolerance)` if you want fewer points, or do lots of other stuff to get a valid non-intersecting linear ring...)